feat: Revamp Script to Follow Better System

Z0y6h0kS9X opened the pull request:
### Description
There are a number of improvements and changes from the original script for usability, maintainability, and performance.

#### Changes
- Moved the variables to a `config.json` for cleaner definition of script inputs.
- Added code comments and documentation inside of the script.
- Renamed the `SYSTEM_TO_FOLDER_MAP` mapping to `config.json`.
- Renamed variables to make them clearer and follow best practice naming conventions more closely.
- Added ability to customize output
  - This is currently limited to generating a full report, or only a list of ROMs without a RetroAchievement match.
- Reworked API Call logic format to make it easily extendable.
- Added additional error checks and handling.
- Added core logic to to functions for easier maintainability and extendibility.
- Added ability to match common names/aliases for platforms to save on manual mapping configuration.
- Removed Write-Progress bars in favor of an inline solution for better reliability and eventual cross-platform compatibility (assuming PowerShell is installed).
- Moved ROM Matching process to a function for eventual individual ROM functionality.
- Updated the README to reflect changes

Commit:
Updating Script to New Process

Z0y6h0kS9X commented:
Hope this helps move it in the direction you want.  Renaming the ROMs is pretty trivial since you are already pulling the Title in the `Get-RAGamesList` function - I'd recommend adding an optional variable in the `config.json` similar to the 'OUTPUT_MISSING_ONLY' in order to specify whether the user wants to do a rename.  If so, we have the Full Path and can just run a `Rename-Item` command on it.

I wanted to get the initial stuff approved before working on your backlog.  The individual ROM one will be a little tricky since it kind of goes against what the script it designed to do, but is possible if you get creative with the enumeration based on the path they supply.

Commit:
fix: removed unused parameters

Siallus commented:
Hey, I appreciate you taking an interest in this project!

I generally like the changes you've suggested. I normally would not opt to have a separate config file for a script of this size, but given that we require the use of an API key, it would make more sense to move it out of the main script. That said, I'd prefer if we moved to a psd1 file instead of json. I tend to like them more for simple use cases since it adheres to PowerShell syntax, allows comments, and does not require escaping paths which may be more confusing to people who aren't familiar with code. The goal of the project is to make this approachable to anyone given the audience.

I have a few nitpicks as well. They're not very important, just personal preferences mostly
- Pascal case for variable names (ex: $baseUrl --> $BaseUrl)
- Spacing inline curly brace code (ex: {$_.Active} --> { $_.Active })
- System.Collections.ArrayList is deprecated. I tend to use System.Collections.Generic.List[Object] instead

I'm not sold on the Mapping_Override functionality. I think allowing for multiple aliases works well enough as-is. Defaulting to the system naming standard that is often used (gba for Game Boy Advance) is nice to have though.

The Show-ProgressBar function works, but I don't see that it's offering anything more than what Write-Progress was doing already, plus we lose the overall system progress indicator. If we want additional console output, we can add "DONE" after each system finishes.


If you wish to continue work on this project, I'm very open to it. 

I had explored ROM renaming before, but the biggest obstacle was around games that were sold with more than 1 version. Pokemon Red and Blue will both return as "Pokémon Red Version | Pokémon Blue Version". There are also concerns with disallowed characters in file names; mostly in Windows. Finding the best universal replacements will probably be annoying. 

Checking individual ROM hashes does go against the premise of the project and it would often just be faster to check an entire library so we can nix that enhancement plan unless you can think of something intuitive.

edit: I forgot to mention too that for the error handling aspect, I think one line write-warning handling would be much preferable than returning entire error blocks. That's something I never got around to adjusting.

edit 2: I also noticed that RAHASHER.exe isn't picked up automatically when the path is provided. The way I have it is far from ideal since it requires that the exe is not renamed, but I don't think it's intuitive at the moment since you'd have to provide the full path (including exe) in the RA_HASHER_PATH config. I'm open to ideas here.

Z0y6h0kS9X commented:
@Siallus 

I'm on mobile sp I can't do a full write up now, but just had an idea for the naming.  Rather than the game title we could do renames to the supported hash title name.  Since we already know the RA ID we can build the url (or maybe they have an API endpoint for it, I'd have to look).

Then we would name the file to what is shown here (depending on the hash you have), it would get around the pokemon red | blue issue you mentioned.  Downside is it may drastically increase the amount of api calls we make unless they have it available as a bulk call like for system game entries, which may cause rate limiting issues.

Normalizing the name isn't difficult, we just need to determine the character replacements and then regex can handle the rest.  

https://retroachievements.org/game/724/hashes

When I get back to the computer I can start updating to match the styling guidelines you want to maintain and start looking into the game hash title.

Siallus commented:
Your suggestion would actually probably work quite well. I never liked how RA prefixes the different types of games with tilde (ie: ~hack~, ~subset~), but the hash names refers to the NoIntro/Redump names which is pretty strong as a standard. That'll be a great feature for those who rip their own roms!

I see your point about the possible API limit, but I think that'll be fine since unlocking achievements uses API calls. I didn't see anything noted in the API docs so I just raised a question in the RA Discord. I'll let you know if I get an answer.

Z0y6h0kS9X commented:
@Siallus 

Looks like this is the API call we can use to get the Game > Hash info

https://api-docs.retroachievements.org/v1/get-game-hashes.html

We would need to supply the Game ID, and extract the MD5 and Title properties.  If we go down this road it is a 1 to 1 request (unfortunately it doesn't look like we can pass multiple), we will want to impose a standard wait time to avoid getting rate limited and keys being blocked.  Yes it is used for unlocks, but we are going to be sending potentially 100's of requests in a very short time, which isn't likely with unlocks so they may start blocking connections.

We may want to consider implementing a cache file system (at least for development) so we aren't pegging their servers constantly for unchanging data, for similar reasons.

Siallus commented:
It looks like there are no hard limits on API usage. They allow a "fair amount" of burst requests per user. 

To be honest though, I'm starting to wonder if this feature is too much scope creep for what this project aims to do. If we were to implement a feature to rename based on file hash, it would be smart to host the data (or find an alternative source) outside of RA.

Z0y6h0kS9X commented:
Finally getting back to a PC so I'll start replying in earnest (partially, it's late so I can only respond to pieces).

"_I'd prefer if we moved to a psd1 file instead of json. I tend to like them more for simple use cases since it adheres to PowerShell syntax, allows comments, and does not require escaping paths which may be more confusing to people who aren't familiar with code._"

 - I see what you mean, I am a big fan of code comments and it is my one gripe with JSON.  I am not super familiar with PSD1 files since they are generally associated with PowerShell modules, so I didn't think it fit.  I could be wrong here and will look into it.

#### Styling Nitpicks
"_Pascal case for variable names (ex: $baseUrl --> $BaseUrl)_"
 - Guess this is just my habits of using Camel Casing, I don't care one way or the other so I can update it no problem.

"_Spacing inline curly brace code (ex: {$.Active} --> { $.Active })_"
 - I agree this looks better, my IDE thinks otherwise so I'll have to manually update them.

_"System.Collections.ArrayList is deprecated. I tend to use System.Collections.Generic.List[Object] instead"_
 - I wasn't aware of the deprecation, I'll change those back over.

_"I'm not sold on the Mapping_Override functionality. I think allowing for multiple aliases works well enough as-is. Defaulting to the system naming standard that is often used (gba for Game Boy Advance) is nice to have though."_
 - I was toying with the idea of removing it and just telling people to add it to the aliases, ultimately I decided to leave it as an option so users would not have to remember the comma - but easily remedied.

#### Show-ProgressBar vs Write-Progress
While I was working on it I noticed some odd behaviors in the write-progress implementation when you passed an empty platform after a platform with ROMs.  So I moved to my own progress indicator to get around it.

#### RA_HASHER_PATH
_"I also noticed that RAHASHER.exe isn't picked up automatically when the path is provided. The way I have it is far from ideal since it requires that the exe is not renamed, but I don't think it's intuitive at the moment since you'd have to provide the full path (including exe) in the RA_HASHER_PATH config. I'm open to ideas here."_
- Yeah I had it explicitly stated so that it didn't matter what the executable was called.  An alternative is that we instruct users to put the executable in the root directory from where the script it run and we execute it.  It would still need to be called 'RAHasher.exe' to avoid any issues there, though this would allow us to remove the configuration option entirely if you are worried about it.  I think the realistic use case of someone renaming RAHasher.exe is pretty limited so I wouldn't stress about it too much.

#### Individual Matching
_"Checking individual ROM hashes does go against the premise of the project and it would often just be faster to check an entire library so we can nix that enhancement plan unless you can think of something intuitive."_
 - The only way that I could see it occurring reasonably would be an overhaul of the script flow entirely.  Rather than relying on a config.json or psd1 file for configuration, we prompt for information and the flow would occur like this.
 
 1. Prompt for RA-User/Api-Key (or maybe leave this as a configurable option) for reuse - though it seems to me we only need the API Key, not the username.
 2. Store response(s) as secure-string objects, or read them in as secure-string objects from config.
 3. Prompt for path to folder/rom
  - We then do a series of checks:
   1. If it is a file, assume it is an individual ROM.  Try to enumerate the platform based on the directory name the rom is in with the Mapping aliases.  If not, prompt for a platform using the results from the Get-RASystems function.  At that point we would have the systemID and the rom file path, we hash it and look it up.
   2. If it is a directory, try to perform the current flow where it tries to determine the platforms.  If it doesn't find any, or there are no subdirectories in the folder, assume it is an individual platform folder and prompt for the System following the step above.  At that point if it is a platform folder rather than a library folder, follow the flow to scan the files and try to get the matches for the roms.
   
This approach would allow for 3 scenarios:
 - Identification of an individual ROM
 - Identification of an individual Platform
 - Identification of an entire library
 
 #### Renaming Out of Scope
 - The core functionality is just for matching the ROM to an RA Hash, so renaming the files is outside of that case.  That said, it is not difficult to implement.  We will need to either host the data elsewhere (stored in the GitHub repo and pulled down?) or perform the matching using RA Hasher for everything, get only valid matches, then do the requests in bursts (of 10 or so) with a 5 second wait timer between batches of requests to avoid too many requests all at once.  What do you think?

Siallus commented:
I'm happy to review the write-progress issues tomorrow when I get back from work. Can you clarify the behavior you're seeing? Is there a bug with file counts, or is it graphical?

Z0y6h0kS9X commented:
The issue lies with how the 2nd step (sub-step) occurs.  In the event that the following directory has no ROMs in it - and/or the folder only has 1 ROM in it, it does not update correctly.

Here I set up an n64 directory with 1 ROM in it (Super Mario 64), it scanned it and moved on to the next directory (NES/Famicom).  You'll note that the sub-step does not clear out and reset, despite the platform being incremented.  furthermore, there is only 1 ROM, so it should display as 100% after parsing it, but it does not.

![image](https://github.com/user-attachments/assets/6ddef9af-6e15-407d-8747-2a83330e8b2b)

Z0y6h0kS9X commented:
Updated the comment response with some more information

https://github.com/Siallus/RetroAchievementsLibraryValidator/pull/1#issuecomment-2791345671

Z0y6h0kS9X commented:
@Siallus 

Don't know if you have had a chance to review my full response above or not.  I fixed the styling to match your preferences but was hopiygor your input before making other changes.

Speaking of which I thought of a functionality that we can do. The report is great but we can go one step further.  What do you think of moving unmatched files to the output directory (inside platform named sub-folders).  This would allow users not only to have the report, but can see the roms per platform that would need to be adjusted.

The downside to the above would be if the rom does not have an RA entry, they would just move it back.  It should be a configurable option that you have to explicitly enable, but there are pros/cons to either.

Siallus commented:
I didn't get a chance to review last night, and unfortunately I'm going to be tied up tonight and most of tomorrow. I do want to get the new flow approved quickly though to allow for more pull requests on those additional features.

Commit:
Updating to incorporate feedback

Commit:
rename script to match original

Z0y6h0kS9X commented:
@Siallus 

I incorporated changes you wanted, except those that we might want to save for future iterations and they can be viewed on the PR.  

### Changes
 - Moved the JSON to a PSD1 file
 - Removed RAUsername, since only the API Key is necessary
 - Removed the MappingOverride field, instead moving to use aliases exclusively
 - Updated README with changes

Siallus commented:
I have a little time this morning to review. Some initial findings:
- Some files that should not have matches, are returning matches to all games of that system:
![image](https://github.com/user-attachments/assets/f25b9261-cf55-47f5-8757-6ac0185e5ac8)

- RAHasher still requires a full path to the exe. I like your suggestion of just pulling this into the script directory which would allow us to remove that config
- I would prefer to stick with Write-Progress. I have a fix coming for that after I post this message. It comes with a bandaid fix for RAHasher calls over non-ROM files, but I'll have another "real" fix for that a bit later.
- All upper function declarations are interesting, I don't think I've ever seen that before. If we can stick with all lower or proper case for all keywords, that would be preferable

I will continue testing and keep you posted. 

I was curious what your background in PowerShell looked like too if you're open to sharing a little about yourself. I see you have some of your own projects with it.

Commit:
Re-Adds Write-Progress and RAHasher function

Commit:
change to GameMatch regex

Z0y6h0kS9X commented:
@Siallus 
_Some files that should not have matches, are returning matches to all games of that system:_
 - Probably due to the regex '!!', I changed it to '-eq' to hopefully correct that.  I have it aborting when I try to create that file and run it.
 
![image](https://github.com/user-attachments/assets/ec1811fa-0a33-4960-b11e-56afbc775385)

_RAHasher still requires a full path to the exe. I like your suggestion of just pulling this into the script directory which would allow us to remove that config_
 - I don't like the idea of embedding a 3rd party executable in the repo, so I introduced a new function that downloads and extracts it if it is not in the current directory.  Currently it is pinned to 1.8.0, but we can rework it to always pull latest if you think there is value in doing that.

_I would prefer to stick with Write-Progress. I have a fix coming for that after I post this message. It comes with a bandaid fix for RAHasher calls over non-ROM files, but I'll have another "real" fix for that a bit later._
 - I removed the Show-ProgressBar function and reintroduced the Write-Progress functionality.

_All upper function declarations are interesting, I don't think I've ever seen that before. If we can stick with all lower or proper case for all keywords, that would be preferable_
- I changed it to Proper casing as that is closer to the Pascal casing standard we have for variables.

### Overview
 - Removed the Show-ProgressBar function
 - Reintroduced Write-Progress for progress indicator
 - Added .gitignore for dev files
 - Added ability to read in dev file(s) without affecting core repo files
 - Updated the FUNCTIONs declarations to Function for consistency
 - Added the Get-RAHasher function to download the RAHasher executable from GitHub if it is not present
 - Reworked and added code comments for clearer documentation


### About Me
I taught myself PowerShell while working in tech support to help with my job.  Since then I wrote many tools and utilities in PowerShell, which eventually landed me a job as a Jr, Software Engineer.  Now my main language is Golang, but I still use PowerShell for rapid prototyping and some personal projects where appropriate - as you have noticed.

Commit:
updated readme to remove RAHasher

Siallus commented:
Maybe we add a config to ask if the user wants the hasher downloaded. I don't like my scripts to download exe's, let alone random Github scripts. Otherwise we would add instructions to the Readme for how to download the program and where to place it. It's better than asking the user for explicit pathing either way.

I'll check the code revisions tonight or tomorrow.

Siallus commented:
I'm gonna start some more testing shortly here.

I see that your hash matching logic is effectively the same as the original, but to be honest I'm not sure why I wrote it that way. It would be far better to use an "-in" condition. I tested it and it seems to work as expected so I'm gonna promote that in a sec.

`$RomMatch = $RAGames | Where-Object { $_.hashes -match $FileHash }` --> `$RomMatch = $RAGames | Where-Object { $FileHash -in $_.hashes }`


Sounds like our career paths a bit similar. I was doing tech support as well, but I didn't end up picking up PS until a few months into a sysadmin role. I tend to overuse it for far more than I should though, even after becoming a data engineer. I'm convinced that it's just about the fastest language for prototyping, or at least, I haven't found anything better.

Siallus commented:
Spent some time looking over the edits and I have some items for review. This list includes the items that I mentioned in my comments above. Moving forward, I'll try to put all my thoughts into a single comment in order to make edits easier. I was able to fix a couple of things that I found so I've included those changes as suggestions:

- RA games that have more than 1 hash will fail to match
	Prior: `$GameMatch = $GameList | Where-Object { $_.hashes -eq $GameHash }`
	Revised: `$GameMatch = $GameList | Where-Object { $GameHash -in $_.hashes }`

- The output csv contains the same System and Path as the initial system and path that are checked
	There are multiple possible fixes, but it appears that the $Platform variable is not initialized before use and is not passed in as an argument so I believe it just never updates

- 'Get-RAHasher' is a good function name, but I'm thinking Install-RAHasher would be more descriptive for this process

- I like the RAHasher download flow, but I feel that we should add a step after the initial check to confirm that the user wants to download. If declined, write-host or write-warning something along the lines of "RAHasher was not downloaded. Exiting script.`n`nYou may manually download RAHasher via the steps in the Github documentation and rerun the script.

- The PlatformList compilation process may be better suited as a function. This would bring the "main" section down to pretty much just error handling, said function call, loops, and output processing

- I'm not a huge fan of for loops in place of foreach, but if the flow works then I'm not against keeping them. If we do stick with them, I think we should do the standard 'i', 'j', 'k' index names in lower case.

- When we get the systemId in the first for loop, we should remove the regex match and put in an eq instead since the "^platform$" logic would require that the platform name be perfect anyhow. Plus, all platform names are normalized already as part of the Platform list compilation which calls Get-PlatformFromName

	Prior: `$SystemId = ($RaSystems | Where-Object { $_.Name -match "^$($PlatformList[$L].Platform)$" }).ID`
	Revised: `$SystemId = ($RaSystems | Where-Object { $_.Name -eq $($PlatformList[$L].Platform) }).ID`

- I feel that we should change "platform" back to "system" to match the RA naming convention. Using "games" for RA endpoint calls and "ROMs" when referencing the files on the user's file system makes sense, but I don't feel that we could make a similar connection for "platform"

- Write-Progress still looks buggy in both our implementations. I think we should implement a hashtable of all supported ROM file types per system as an additional property in the PlatformMapping config. Once implemented, before the RAHasher check of a given ROM, we should check to see if the file has the correct extension for the current system in the loop before its parsed. If its still wonky then we can return to your function. This can be handled at a later date, I'm happy to implement it once this pr is accepted.

Z0y6h0kS9X commented:
Sorry, been away from the computer for a bit, I'll review your comments and see if I can update the script to incorporate the changes at some point this weekend.

Siallus commented:
No worries, and no rush! I hadn't intended for collaboration on this project so I never made any style or maintenance guidelines, yet I tend to be picky about that stuff. Apologies for that.

I realize we haven't made much progress which may be disheartening. We could always break this PR up into smaller components to make troubleshooting and implementation easier.
